File: kata-8-kyu/powers-of-2/python/powers_of_2_test.py
# + =====================================================================================================================
# +
# + 8 kyu - Powers of 2  [ ID: 57a083a57cb1f31db7000028 ] (powers-of-2)
# + URL: https://www.codewars.com/kata/57a083a57cb1f31db7000028
# + Category: REFERENCE  |  Tags: MATHEMATICS | FUNDAMENTALS
# +
# + =====================================================================================================================

import codewars_test as test
from powers_of_2 import powers_of_two


@test.describe("Fixed Tests")
def fixed_tests():
    @test.it("Basic Test Cases")
    def basic_test_cases():
        test.assert_equals(powers_of_two(0), [1])
        test.assert_equals(powers_of_two(1), [1, 2])
        test.assert_equals(powers_of_two(4), [1, 2, 4, 8, 16])


@test.describe("Random Tests")
def random_test():
    from random import shuffle

    tests = list(range(0, 201))
    shuffle(tests)

    def _powers_of_two(n):
        return [2**x for x in range(0, n + 1)]

    for n in tests:

        @test.it(f"testing for powers_of_two({n})")
        def test_case():
            test.assert_equals(powers_of_two(n), _powers_of_two(n))

File: kata-8-kyu/powers-of-2/javascript/powersOf2.Test.js
//+ ====================================================================================================================
//+
//+ 8 kyu - Powers of 2  [ ID: 57a083a57cb1f31db7000028 ] (powers-of-2)
//+ URL: https://www.codewars.com/kata/57a083a57cb1f31db7000028
//+ Category: REFERENCE  |  Tags: MATHEMATICS | FUNDAMENTALS
//+
//+ ====================================================================================================================

import { assert, vi } from "vitest"
const { powersOfTwo } = require("./powersOf2")
const _shuffle = require("lodash").shuffle

vi.setConfig.truncateThreshold = 0

describe("Basic Tests", function () {
  it("Testing for fixed tests", () => {
    assert.deepEqual(powersOfTwo(0), [1])
    assert.deepEqual(powersOfTwo(1), [1, 2])
    assert.deepEqual(powersOfTwo(4), [1, 2, 4, 8, 16])
  })
})

describe("Random Tests", function () {
  const solution = (n) => [...Array(n + 1)].map((x, y) => Math.pow(2, y))
  const shuffle = _shuffle([...Array(201)].map((x, y) => y))

  it("Testing for 201 random tests", () => {
    for (const n in shuffle) {
      assert.deepEqual(powersOfTwo(shuffle[n]), solution(shuffle[n]), `Testing for n = ${n}`)
    }
  })
})
